File: main/src/permissions.rs
//! Permission modes and approval policy.
//!
//! Mode semantics live in the C core (`csrc/permissions.c`) so every caller
//! — interactive sessions, one-shot prompts, apply mode, counsel mode, and
//! tests — shares one decision table. This module provides the Rust types,
//! clap/serde names with legacy aliases, and the interactive approval prompt.

use clap::ValueEnum;
use serde::{Deserialize, Serialize};

/// Canonical CLI/serde mode names: `auto-approve`, `all-approve`,
/// `manual-approve`, plus legacy `counsel` and `file-only`. The legacy
/// names `auto`, `allow`, and `request-permission` remain accepted aliases.
#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize, ValueEnum)]
#[serde(rename_all = "kebab-case")]
pub enum Mode {
    #[default]
    #[value(name = "auto-approve", alias = "auto")]
    #[serde(rename = "auto-approve", alias = "auto")]
    Auto,
    Counsel,
    #[value(name = "all-approve", alias = "allow")]
    #[serde(rename = "all-approve", alias = "allow")]
    Allow,
    #[value(name = "manual-approve", alias = "request-permission")]
    #[serde(rename = "manual-approve", alias = "request-permission")]
    RequestPermission,
    FileOnly,
}

impl Mode {
    /// Mode value in the C core decision table.
    pub fn code(self) -> i32 {
        match self {
            Self::Auto => crate::core::MODE_AUTO_APPROVE,
            Self::Counsel => crate::core::MODE_COUNSEL,
            Self::Allow => crate::core::MODE_ALL_APPROVE,
            Self::RequestPermission => crate::core::MODE_MANUAL_APPROVE,
            Self::FileOnly => crate::core::MODE_FILE_ONLY,
        }
    }

    pub fn from_code(code: i32) -> Option<Self> {
        match code {
            crate::core::MODE_AUTO_APPROVE => Some(Self::Auto),
            crate::core::MODE_COUNSEL => Some(Self::Counsel),
            crate::core::MODE_ALL_APPROVE => Some(Self::Allow),
            crate::core::MODE_MANUAL_APPROVE => Some(Self::RequestPermission),
            crate::core::MODE_FILE_ONLY => Some(Self::FileOnly),
            _ => None,
        }
    }

    /// Canonical CLI name, from the C core.
    pub fn as_str(self) -> &'static str {
        crate::core::mode_canonical_name(self.code()).unwrap_or("auto-approve")
    }

    /// Shift+Tab cycle order, from the C core.
    pub fn next(self) -> Self {
        Self::from_code(crate::core::mode_next(self.code())).unwrap_or(Self::Auto)
    }

    /// Parse a canonical or legacy alias name through the C core.
    pub fn parse(name: &str) -> Option<Self> {
        crate::core::mode_parse(name).and_then(Self::from_code)
    }

    pub fn description(self) -> &'static str {
        crate::core::mode_description(self.code()).unwrap_or("")
    }
}

/// Approval is explicit and fail-closed, including redirected stdin and EOF.
/// Prompts use plain language ("Cntx wants to: ...") so non-technical users
/// can decide without reading raw tool JSON.
pub fn confirm(action: &str) -> bool {
    use std::io::{self, IsTerminal, Write};
    if !io::stdin().is_terminal() {
        eprintln!(
            "Cntx needs your permission to {action}, but this is not an interactive terminal. Rerun interactively, or use --mode all-approve to allow permitted tools without prompting."
        );
        return false;
    }
    eprintln!("Cntx wants to: {action}");
    eprint!("Allow once? [y = yes / n = no] ");
    let _ = io::stderr().flush();
    let mut answer = String::new();
    io::stdin().read_line(&mut answer).is_ok()
        && matches!(answer.trim().to_ascii_lowercase().as_str(), "y" | "yes")
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Operation {
    ReadFile,
    WriteFile,
    Shell,
    Network,
}

impl Operation {
    fn code(self) -> i32 {
        match self {
            Self::ReadFile => crate::core::OP_READ,
            Self::WriteFile => crate::core::OP_WRITE,
            Self::Shell => crate::core::OP_SHELL,
            Self::Network => crate::core::OP_NETWORK,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PermissionDecision {
    Allow,
    Ask,
    Deny,
}

impl PermissionDecision {
    fn from_code(code: i32) -> Self {
        match code {
            crate::core::DECISION_ASK => Self::Ask,
            crate::core::DECISION_DENY => Self::Deny,
            _ => Self::Allow,
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct PermissionPolicy {
    mode: Mode,
}

impl PermissionPolicy {
    pub fn new(mode: Mode) -> Self {
        Self { mode }
    }

    pub fn mode(&self) -> Mode {
        self.mode
    }

    /// Delegate the decision to the C core. This is the single policy
    /// implementation shared by the tool loop, apply mode, and counsel.
    pub fn decide(&self, operation: Operation) -> PermissionDecision {
        PermissionDecision::from_code(crate::core::permission_decide(
            self.mode.code(),
            operation.code(),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn canonical_names_come_from_c_core() {
        assert_eq!(Mode::Auto.as_str(), "auto-approve");
        assert_eq!(Mode::Allow.as_str(), "all-approve");
        assert_eq!(Mode::RequestPermission.as_str(), "manual-approve");
        assert_eq!(Mode::Counsel.as_str(), "counsel");
        assert_eq!(Mode::FileOnly.as_str(), "file-only");
    }

    #[test]
    fn legacy_aliases_resolve() {
        assert_eq!(Mode::parse("auto"), Some(Mode::Auto));
        assert_eq!(Mode::parse("allow"), Some(Mode::Allow));
        assert_eq!(
            Mode::parse("request-permission"),
            Some(Mode::RequestPermission)
        );
        assert_eq!(Mode::parse("auto-approve"), Some(Mode::Auto));
        assert_eq!(Mode::parse("all-approve"), Some(Mode::Allow));
        assert_eq!(Mode::parse("manual-approve"), Some(Mode::RequestPermission));
    }

    #[test]
    fn cycle_matches_contract() {
        assert_eq!(Mode::Auto.next(), Mode::Allow);
        assert_eq!(Mode::Allow.next(), Mode::RequestPermission);
        assert_eq!(Mode::RequestPermission.next(), Mode::Auto);
        assert_eq!(Mode::Counsel.next(), Mode::Auto);
        assert_eq!(Mode::FileOnly.next(), Mode::Auto);
    }

    #[test]
    fn policy_delegates_to_c() {
        let policy = PermissionPolicy::new(Mode::Auto);
        assert_eq!(
            policy.decide(Operation::ReadFile),
            PermissionDecision::Allow
        );
        assert_eq!(policy.decide(Operation::Shell), PermissionDecision::Ask);
        let file_only = PermissionPolicy::new(Mode::FileOnly);
        assert_eq!(file_only.decide(Operation::Shell), PermissionDecision::Deny);
    }
}

File: main/csrc/permissions.c
/* Approval mode semantics: the single source of truth for permission
 * decisions across interactive sessions, one-shot prompts, apply mode, and
 * counsel mode. */
#include "cntx.h"

#include <string.h>

/* The exact decision table from the product contract:
 *
 * | Operation                    | auto | all | manual | file-only | counsel |
 * | Explicit read/glob/grep tool | Allow| Allow| Ask   | Allow     | Allow   |
 * | In-root write/edit/apply     | Ask  | Allow| Ask   | Allow     | Ask     |
 * | Shell command                | Ask  | Allow| Ask   | Deny      | Ask     |
 * | Outside-root direct write    | Deny | Deny | Deny  | Deny      | Deny    |
 *
 * Outside-root denial is layered on top of the mode decision by the sandbox
 * (containment check in Rust); this table only expresses the mode policy.
 */
cntx_decision_t cntx_permission_decide(int mode, int operation) {
    if (mode < CNTX_MODE_AUTO_APPROVE || mode > CNTX_MODE_FILE_ONLY ||
        operation < CNTX_OP_READ || operation > CNTX_OP_NETWORK) {
        return CNTX_DECISION_DENY;
    }

    switch (mode) {
    case CNTX_MODE_ALL_APPROVE:
        return CNTX_DECISION_ALLOW;
    case CNTX_MODE_MANUAL_APPROVE:
        return CNTX_DECISION_ASK;
    case CNTX_MODE_FILE_ONLY:
        switch (operation) {
        case CNTX_OP_READ:
        case CNTX_OP_WRITE:
            return CNTX_DECISION_ALLOW;
        default: /* shell, network */
            return CNTX_DECISION_DENY;
        }
    case CNTX_MODE_AUTO_APPROVE:
    case CNTX_MODE_COUNSEL:
    default:
        switch (operation) {
        case CNTX_OP_READ:
            return CNTX_DECISION_ALLOW;
        default: /* write, shell, network */
            return CNTX_DECISION_ASK;
        }
    }
}

const char *cntx_mode_canonical_name(int mode) {
    switch (mode) {
    case CNTX_MODE_AUTO_APPROVE:
        return "auto-approve";
    case CNTX_MODE_COUNSEL:
        return "counsel";
    case CNTX_MODE_ALL_APPROVE:
        return "all-approve";
    case CNTX_MODE_MANUAL_APPROVE:
        return "manual-approve";
    case CNTX_MODE_FILE_ONLY:
        return "file-only";
    default:
        return NULL;
    }
}

const char *cntx_mode_description(int mode) {
    switch (mode) {
    case CNTX_MODE_AUTO_APPROVE:
        return "allow reads and require approval for writes or shell actions";
    case CNTX_MODE_COUNSEL:
        return "use token-efficient model counsel; same approval behavior as auto-approve";
    case CNTX_MODE_ALL_APPROVE:
        return "execute permitted tools without prompts; file containment still applies";
    case CNTX_MODE_MANUAL_APPROVE:
        return "ask before every tool, file, or shell operation";
    case CNTX_MODE_FILE_ONLY:
        return "allow file reads/writes, but block shell and network";
    default:
        return NULL;
    }
}

int cntx_mode_parse(const char *name, int *out_mode) {
    if (name == NULL || out_mode == NULL) {
        return CNTX_ERR_INVALID_ARGUMENT;
    }
    if (strcmp(name, "auto-approve") == 0 || strcmp(name, "auto") == 0) {
        *out_mode = CNTX_MODE_AUTO_APPROVE;
    } else if (strcmp(name, "counsel") == 0) {
        *out_mode = CNTX_MODE_COUNSEL;
    } else if (strcmp(name, "all-approve") == 0 || strcmp(name, "allow") == 0) {
        *out_mode = CNTX_MODE_ALL_APPROVE;
    } else if (strcmp(name, "manual-approve") == 0 ||
               strcmp(name, "request-permission") == 0) {
        *out_mode = CNTX_MODE_MANUAL_APPROVE;
    } else if (strcmp(name, "file-only") == 0) {
        *out_mode = CNTX_MODE_FILE_ONLY;
    } else {
        return CNTX_ERR_NOT_FOUND;
    }
    return CNTX_OK;
}

int cntx_mode_next(int mode) {
    switch (mode) {
    case CNTX_MODE_AUTO_APPROVE:
        return CNTX_MODE_ALL_APPROVE;
    case CNTX_MODE_ALL_APPROVE:
        return CNTX_MODE_MANUAL_APPROVE;
    case CNTX_MODE_MANUAL_APPROVE:
        return CNTX_MODE_AUTO_APPROVE;
    /* Legacy extra modes cycle back to the canonical default. */
    case CNTX_MODE_COUNSEL:
    case CNTX_MODE_FILE_ONLY:
    default:
        return CNTX_MODE_AUTO_APPROVE;
    }
}
